### PR TITLE
Add support for FMT_STRING compile time checking.

### PR DESCRIPTION
**Pull request by Tridacnid**
Additionally, add negative compilation unit test for compilers and c++ standard that support relaxed constexpr.

Hello, I discovered that the FMT_STRING compile time checking wasn't supported and figured I wanted it enough to add it. After the fact I found #1369, hopefully this pull request addresses all concerns by the maintainers. If not, please let me know what I can do to improve the design to make the pull more appealing.

I'm a Linux developer but I did my best to check the Windows support by booting to Windows10 and running the unit tests in Visual Studio with `-DSPDLOG_BUILD_TESTS=ON -DSPDLOG_BUILD_WARNINGS=ON -DSPDLOG_WCHAR_SUPPORT=ON`

### CODE REVIEW

**Comment by Tridacnid** [comment]
I changed quite a few things to perfect forwarding.
```
    template<typename T>
    void trace(T &&msg)
    {
        log(level::trace, std::forward<T>(msg));
    }
```

If this is not desirable, let me know and I can switch them back.

File: include/spdlog/logger.h [hunk 1]
```diff
@@ -73,78 +73,74 @@ class SPDLOG_API logger
...
+    void log(level::level_enum lvl, T &&msg)
     {
-        log(source_loc{}, lvl, msg);
+        log(source_loc{}, lvl, std::forward<T>(msg));
     }
 
     // T can be statically converted to string_view
-    template<class T, typename std::enable_if<std::is_convertible<const T &, spdlog::string_view_t>::value, T>::type * = nullptr>
-    void log(source_loc loc, level::level_enum lvl, const T &msg)
+    template<class T, typename std::enable_if<std::is_convertible<T &&, spdlog::string_view_t>::value, T>::type * = nullptr>
```
Comment (Tridacnid) [inline include/spdlog/logger.h:140]: SFINAE + universal reference?

File: include/spdlog/logger.h [hunk 2]
```diff
@@ -179,48 +175,48 @@ class SPDLOG_API logger
     }
 
     // T cannot be statically converted to string_view or wstring_view
-    template<class T, typename std::enable_if<!std::is_convertible<const T &, spdlog::string_view_t>::value &&
-                                                  !is_convertible_to_wstring_view<const T &>::value,
-                          T>::type * = nullptr>
-    void log(source_loc loc, level::level_enum lvl, const T &msg)
+    template<class T,
+        typename std::enable_if<!std::is_convertible<T &&, spdlog::string_view_t>::value && !is_convertible_to_wstring_view<T &&>::value,
```
Comment (Tridacnid) [inline include/spdlog/logger.h:179]: SFINAE + universal reference?

File: include/spdlog/logger.h [hunk 3]
```diff
@@ -251,51 +247,9 @@ class SPDLOG_API logger
...
-    template<typename... Args>
-    void critical(wstring_view_t fmt, const Args &... args)
-    {
-        log(level::critical, fmt, args...);
-    }
-
     // T can be statically converted to wstring_view
-    template<class T, typename std::enable_if<is_convertible_to_wstring_view<const T &>::value, T>::type * = nullptr>
-    void log(source_loc loc, level::level_enum lvl, const T &msg)
+    template<class T, typename std::enable_if<is_convertible_to_wstring_view<T &&>::value, T>::type * = nullptr>
```
Comment (Tridacnid) [inline include/spdlog/logger.h:251]: SFINAE + universal reference?

**Comment by Tridacnid** [comment]
I have marked a couple spots I'm unsure about, I've never combined universal references and SFINAE so I'm not sure my change is 100% correct in these places.

**Comment by gabime** [comment]
> If this is not desirable, let me know and I can switch them back.

Probably not needed. Lets concentrate on the FMT_STRING and remove other non related changes. universal references are very confusing and i prefer to avoid them if not really needed.

**Comment by Tridacnid** [comment]
> > If this is not desirable, let me know and I can switch them back.
> 
> Probably not needed. Lets concentrate on the FMT_STRING and remove other non related changes. universal references are very confusing and i prefer to avoid them if not really needed.

In hindsight,
```
    template<typename T>
    void trace(T &&msg)
    {
        log(level::trace, std::forward<T>(msg));
    }
```
was a bad example, because we'd want to catch 
```
spdlog::trace(FMT_STRING("I have too many expected params {}!"));
```

**Comment by Tridacnid** [comment]
> > If this is not desirable, let me know and I can switch them back.
> 
> Probably not needed. Lets concentrate on the FMT_STRING and remove other non related changes. universal references are very confusing and i prefer to avoid them if not really needed.

Can do. If avoiding universal references is a goal I can probably remove them everywhere, including the templated FormatString type I added.

**Comment by Tridacnid** [comment]
I'd like to update the CI to run the negative compilation test when possible (clang10, gcc7). Is something like this acceptable?

```
    - env: GCC_VERSION=7 BUILD_TYPE=Release CPP=14
      os: linux
      addons: *gcc7

# further down, in script
script:
    ...
    -DSPDLOG_BUILD_FAILING_TESTS=ON \ #Activates negative test when C++ version >11 and compiler supports relaxed constexpr
    ...
```

**Comment by gabime** [comment]
spdlog is c++11 so lets not. Also, those tests are probably more relevant to fmt than to spdlog.

**Comment by Tridacnid** [comment]
> spdlog is c++11 so lets not. Also, those tests are probably more relevant to fmt than to spdlog.

If that's the direction you'd like to take I'll defer to you. I added it so that a regression in the API could be detected when using FMT_STRING.

**Comment by Tridacnid** [comment]
Do you even want the FMT_STRING compilation error test? If it isn't going to run with CI I worry it would fall out of maintenance and become a burden.

**Comment by gabime** [comment]
Please remove the test. why c++14 btw? This feature wont work in c++11?

**Comment by Tridacnid** [comment]
FMT_STRING needs relaxed constexpr to work correctly. In c++11 it falls back to being a runtime check.

**Comment by gabime** [comment]
Ok. so after removing the universal refs and tests we this would be much smaller pr with minimal changes 2 fo 2 files only: logger.h  spdlog.h. Waiting for your fixes

**Comment by Tridacnid** [comment]
Sounds like a plan, I'll try to get the updates pushed in the next day or two.

**Comment by Tridacnid** [comment]
Updated.

**Comment by gabime** [comment]
Thanks. Why the universal ref for FormatString?

**Comment by Tridacnid** [comment]
I debated with myself whether you wanted me to eliminate those as well, sounds like I should have.

No reason other than that's how I originally implemented it, I'll run it to ground this evening.

**Comment by Tridacnid** [comment]
Updated.

**Comment by gabime** [comment]
Merged. Thanks @Tridacnid

**Pull request merged**